### PR TITLE
[ART-8100] Refactor searchParams processing in getBuilds

### PR DESCRIPTION
**Pull request by MartinPl0**
Enhanced the clarity and functionality of the getBuilds function:
- Introduced a clear check for whether a value is enclosed in quotes.
- Grouped logic for exact and partial matches based on keys, improving readability.
- Added comments for better code understanding.

### CODE REVIEW

File: components/api_calls/build_calls.js [hunk 1]
```diff
@@ -6,14 +6,18 @@ export async function getBuilds(searchParams) {
     for (const key in searchParams) {
         let value = searchParams[key];
 
-        if ((value[0] === "\"" && value[value.length - 1] === "\"") || (value[0] === "'" && value[value.length - 1] === "'")) {
-            value = value.substring(1, value.length - 1);
+        // Check if the value is enclosed in double or single quotes.
+        const isEnclosedInQuotes = (value[0] === "\"" && value[value.length - 1] === "\"") || (value[0] === "'" && value[value.length - 1] === "'");
```
Comment (ashwindasr) [inline components/api_calls/build_calls.js:10]: Hmm couldn't a regex be an easier way to check?

File: components/api_calls/build_calls.js [hunk 2]
```diff
@@ -6,14 +6,18 @@ export async function getBuilds(searchParams) {
...
+        
+        if (isEnclosedInQuotes) {
+            value = value.substring(1, value.length - 1); // Strip the quotes
+            params[key] = value; // Exact match
+        } else {
+            if (key !== "time_iso" && key !== "page" && key !== "brew_task_state")
+                params[`${key}__icontains`] = value; // Partial match
+            else
+                params[key] = value; // Exact match for the excepted keys
```
Comment (ashwindasr) [inline components/api_calls/build_calls.js:19]: So basically this is saying exact match for `time_iso`, `page` and `brew_task_state` right? So isn't a better way to say:

```
if (key === "time_iso" && key === "page" && key === "brew_task_state")
    params[key] = value;
else
    params[`${key}__icontains`] = value;
```
which is easier to understand. Unless you do not think there is no better way than this? :wink: _(Problems of using double negatives)_

**Review by ashwindasr — CHANGES_REQUESTED** [review]

File: components/api_calls/build_calls.js (hunk 2 above)
Comment (MartinPl0) [reply components/api_calls/build_calls.js:19]: The variable key can hold only one value at a time. In the statement:
`if (key === "time_iso" && key === "page" && key === "brew_task_state")`
The && operator (logical AND) checks if all conditions are true. For the entire statement to be true, key would have to be equal to "time_iso", "page", and "brew_task_state" at the same time, which is not possible.

Instead, using `OR ('||')` will correctly check if 'key' matches any of the three values.
I've made the necessary adjustment. Thanks for pointing it out!

File: components/api_calls/build_calls.js (hunk 1 above)
Comment (MartinPl0) [reply components/api_calls/build_calls.js:10]: right! my bad

File: components/api_calls/build_calls.js [hunk 3]
```diff
@@ -6,14 +6,18 @@ export async function getBuilds(searchParams) {
     for (const key in searchParams) {
         let value = searchParams[key];
 
-        if ((value[0] === "\"" && value[value.length - 1] === "\"") || (value[0] === "'" && value[value.length - 1] === "'")) {
-            value = value.substring(1, value.length - 1);
+        // Check if the value is enclosed in double or single quotes using regex.
+        const isEnclosedInQuotes = /^["'].*["']$/.test(value);
```
Comment (ashwindasr) [inline components/api_calls/build_calls.js:10]: So `.*` is something that is best avoided, as a good practice. We will probably only have alpha numerals in addition to `-`. I think taking some time to think about which all values a variable can have or match is always worth it.

**Review by ashwindasr — CHANGES_REQUESTED** [review]

File: components/api_calls/build_calls.js (hunk 2 above)
Comment (ashwindasr) [reply components/api_calls/build_calls.js]: yeah my bad, I wasn't looking at the logical operators. Good catch.

File: components/api_calls/build_calls.js (hunk 3 above)
Comment (MartinPl0) [reply components/api_calls/build_calls.js]: ohh I had no idea, it makes sense thanks for the advice.

**Comment by ashwindasr** [comment]
Redeployed to your hackspace?

**Comment by ashwindasr** [comment]
Feel free to add approve and merge once you've confirmed that its working

**Pull request merged**